### PR TITLE
feat: Add thread information to File I/O spans

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Added a new `data` item to File I/O automatic spans that indicates whether it happens in the main thread.
If it does, we also include the stack trace.

## :bulb: Motivation and Context

See https://github.com/getsentry/rfcs/blob/main/text/0036-auto-instrumentation-ui-thread.md

## :green_heart: How did you test it?

Unit tests for the changes regarding Span creation, to make sure proper data is attached to the span.

We have UI Tests that run the SDK for both, the new code when reading file in the main thread, as the old symbolication steps when capturing errors or messages. This way we know whether the changes in the symbolication process will crash the app or not.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [inline Sources/Sentry/SentryStacktraceBuilder.m:79]: It looks like `retrieveStacktraceFromCursor` is not used during crashes. Maybe we could change it to use `dladdr`, remove this new method, and keep only one method that is faster, because 'retrieveStacktraceFromCursor' run for every event.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:197]: `m`: I would prefer using the `SentryDependencyContainer` or even better pass in the `SentryThreadInspector` as a dependency to `SentryNSDataTracker`, remove the `sharedInstance` from `SentryNSDataTracker` and add the `SentryNSDataTracker` to the `SentryDependencyContainer`, and `SentryNSDataSwizzling` uses the `SentryDependencyContainer` instead of the `SentryNSDataTracker.sharedInstance`.

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (philipphofmann) [reply Sources/Sentry/SentryStacktraceBuilder.m:79]: Yes, please let's do that. That's an excellent idea, but we need to validate this properly.

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (philipphofmann) [inline Sources/Sentry/SentryStacktraceBuilder.m:84]: `h`: Why don't we use `sentrycrashsymbolicator_symbolicate` here? The code looks almost the same and it has support for async markers.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:200]: I love the small optimization `initWithCapacity:` 💯 😀

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:202]: `m`: I'm a bit worried about this piece of code. If you have many file IO operations on the main thread, this can drastically slow you down and increase the payload. We could, for example, only include the 20 topmost frames, which should be enough for you to locate the problem and enough for grouping. I started a thread in [Slack](https://sentry.slack.com/archives/G01PW6HRQNN/p1672825886909859) to discuss thi.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:191]: `m`: I guess you already synced `blocked_main_thread` and `call_stack` with other teams and backend?

**Review by philipphofmann — COMMENTED** [review]
I didn't do a full review yet, as we have some comments that can lead to a bit of refactoring in this PR.

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: Because is super slow.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:191]: This is from the RFC

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (philipphofmann) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: Yes, but only because it calls into `sentrycrashdl_dladdr`. What about passing in a flag to `sentrycrashsymbolicator_symbolicate` to control whether to call `sentrycrashdl_dladdr` or `dladdr` or refactoring it to use the same code?

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:202]: The original solution was to include only frames from the app itself. An idea that I really like. I thought this would not be possible because I would have to remove local symbolization. Now that I refactor the symbolization process and is fast enough I think we can go back to this original idea

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: Got it. I'll take a look whether this is possible,

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [reply Sources/Sentry/SentryNSDataTracker.m:202]: Let's not do this for now, as we only serialize if you are doing file IO on the main thread, and it's acceptable to slow down applications a bit to notify them that they are doing it on the main thread.

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: If you dont mind I will do this refactoring in a follow up PR. We need this feature to be available with v8 if possible.

Comment (philipphofmann) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: When you open an issue so we don't forget about it. It's acceptable.

File: Sources/Sentry/include/SentryStacktraceBuilder.h
Comment (philipphofmann) [inline Sources/Sentry/include/SentryStacktraceBuilder.h:35]: `m`: I'm a bit confused by the name of this function. Why natively? Furthermore, we always pass `YES` to this method, so we can remove the extra param.
```suggestion
- (SentryStacktrace *)buildStacktraceForCurrentThreadNatively;
```

File: Sources/Sentry/include/SentryTracer.h
Comment (philipphofmann) [inline Sources/Sentry/include/SentryTracer.h:109]: `m`: I think we can remove this property. Why don't we check if data contains `@"call_stack"` and then add the debug meta. I'm not a huge fan of setting boolean flags for statuses that can already be checked depending on some existing information.

**Review by philipphofmann — COMMENTED** [review]
First pass, I will give this another one in the afternoon.

File: Sources/Sentry/include/SentryStacktraceBuilder.h
Comment (brustolin) [reply Sources/Sentry/include/SentryStacktraceBuilder.h:35]: Because we are using OS native function to symbolicated instead of using the hand made function. Im open for suggestions on naming ;)

File: Sources/Sentry/include/SentryTracer.h
Comment (brustolin) [reply Sources/Sentry/include/SentryTracer.h:109]: What if the user created a call_stack data themself?

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m:84]: [Issue](https://github.com/getsentry/sentry-cocoa/issues/2602)

File: Sources/Sentry/SentryDebugImageProvider.m
Comment (philipphofmann) [inline Sources/Sentry/SentryDebugImageProvider.m:42]: `l`:
```suggestion
    NSMutableArray<SentryDebugMeta *> *result = [NSMutableArray array];
```

File: Sources/Sentry/SentryDebugImageProvider.m
Comment (philipphofmann) [inline Sources/Sentry/SentryDebugImageProvider.m:47]: `m`: When using an `NSSet` for `addresses` we improve this from `n*m` to `n`.

File: Sources/Sentry/SentryFrameRemover.m
Comment (philipphofmann) [inline Sources/Sentry/SentryFrameRemover.m:13]: 💯 for this change

Comment (philipphofmann) [inline Sources/Sentry/SentryFrameRemover.m:13]: `m`: Please add tests for this change.

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:14]: `m`: Please pass the process info wrapper in as a dependency to `SentryNSDataTracker:init` instead, so the initializer clearly communicates its dependencies.

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:33]: `m`: That's a testing smell. We should remove the `sharedInstance` and move the singleton to the `SentryDependencyContainer`, so we don't have to clean up global state here.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:219]: `h`: This change significantly impacts the payload size. Can't we only send this if the span blocked the main thread? Why do we need to add this data value if it didn't block the main thread?

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:240]: `m`: In that case, I wouldn't set anything. I would only set `blocked_main_thread` if the span blocked the main thread and the frames contain non-system frames.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:236]: `m`: Shouldn't this be?
```suggestion
    if (frames.count == 0) {
```

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:233]: `m`: I don't see any tests for this logic, or am I missing something?

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (philipphofmann) [inline Sources/Sentry/SentryStacktraceBuilder.m:84]: `m`: I think we should check if `dladdr` returns a non-zero value
```suggestion
                            if (dladdr((const void *)cursor->stackEntry.address, &info) != 0) {
```

> On success, these functions return a nonzero value. - https://man7.org/linux/man-pages/man3/dladdr.3.html

Comment (philipphofmann) [inline Sources/Sentry/SentryStacktraceBuilder.m:84]: `h`: `sentrycrashdl_dladdr` uses `CALL_INSTRUCTION_FROM_RETURN_ADDRESS`. Why don't we use it here?

File: Sources/Sentry/SentryTracer.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:723]: `m`: As the tracer already contains so much code, what do you think about extracting some of that code into another class? Something like
```ObjC
if (self.requiresDebugMeta) {
    transaction.debugMeta = [transactionDebugMetaProvider getDebugMetaForSpan:self.rootSpan];
}
```

File: Sources/Sentry/SentryTracer.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:728]:
```suggestion
```

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [inline Sources/Sentry/SentryNSDataTracker.m:232]: `m`: Instead of serializing the frame already here we could also add private property of `NSArray<SentryFrame *>*>` to `SentrySpan` and serialize in when calling `SentrySpan serialize`. Then also retrieving the debugImages in `extractFramesImageAddressFromSpan` would be easier cause we don't have to look into a NSDictionary and do a bunch of casting.

File: Sources/Sentry/SentryTracer.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:695]: `m`: I don't see any tests for this new code, or am I missing something?

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:273]: `m`: This assertion doesn't validate much. I think it would be better to use a `TestSentryThreadInspector` and assert that `SentryNSDataTracker` properly assigns the stacktrace of the `TestSentryThreadInspector`. We could also validate then that it skips system frames.

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:112]: `m`: Shouldn't this teat also call `assertDataSpan`?

**Review by philipphofmann — COMMENTED** [review]
There are many issues to address in this PR before we merge this. I'm looking forward to having this feature in our SDK 😀

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:14]: SentryNSDataTracker is used by `SentryNSDataSwizzling`, and since swizzling is permanent, the reference to `SentryNSDataTracker` will be kept by the swizzled method. 
Thats why `SentryNSDataTracker` and so many other tracker classes used with swizzling are Singletons. 
It is possible to change this approach, but imo not worth it.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:219]: Is just a boolean. How this `significantly impacts the payload size`?
I guess we could not include it, but this is how was decided in the RFC and how android do it.

@romtsn thoughts?

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:232]: Yeah, Im a bit worried about adding extra properties to SentrySpan for every new feature, but I kind liked the idea.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:236]: No. It always have 'main' frame

File: Sources/Sentry/SentryNSDataTracker.m
Comment (brustolin) [reply Sources/Sentry/SentryNSDataTracker.m:240]: Same as https://github.com/getsentry/sentry-cocoa/pull/2573/files#r1071058845

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:33]: Discussion about singleton or not singleton in here https://github.com/getsentry/sentry-cocoa/pull/2573/files#r1071053357

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:112]: Not really necessary

File: Sources/Sentry/SentryNSDataTracker.m
Comment (romtsn) [reply Sources/Sentry/SentryNSDataTracker.m:219]: imo it's fine to always send the boolean, I don't think the payload size will suffer much from this (we also agreed with backend to always send it). More important is that we should not send the stacktrace if the boolean is `false`, but this is already considered on the line 221.

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:14]: Even tho we would just changing the singleton behaviour around, maybe we should only use singleton with the swizzling classes, this classes are already only tested during integration tests.

Comment (philipphofmann) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift:14]: > and since swizzling is permanent, the reference to SentryNSDataTracker will be kept by the swizzled method.

That's already the case.

> Thats why SentryNSDataTracker and so many other tracker classes used with swizzling are Singletons.

They shouldn't know that they are singletons. Instead, the SentryDependencyContainer or another class should decide that for them. Dependency injection is key for testability and readability cause it helps you to understand which dependencies a class has. Now I need to read the code of SentryNSDataTracker and the code of its test to understand how to test it and which internal dependencies it has. IMO it's worth the refactoring.

File: Sources/Sentry/SentryDebugImageProvider.m
Comment (philipphofmann) [reply Sources/Sentry/SentryDebugImageProvider.m:47]: Why did you mark that as resolved when the issue is not addressed?

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [reply Sources/Sentry/SentryNSDataTracker.m:219]: > I don't think the payload size will suffer much from this

For every file I/O span, we now will send `blocked_main_thread`: `false`. Such small things all add up to a payload size increase. IMO the default should be `false`, and we should only send it if `blocked_main_thread`: `true`.

File: Sources/Sentry/SentryNSDataTracker.m
Comment (philipphofmann) [reply Sources/Sentry/SentryNSDataTracker.m:236]: Please add this info as a comment.

**Review by philipphofmann — COMMENTED** [review]
I think a review from a native dev for the small part of C code would be great as suggested by the PR template:

https://github.com/getsentry/sentry-cocoa/blob/56ec5d04c9b2d9c989dfc5e55f904c537bdc2956/.github/pull_request_template.md?plain=1#L19

File: Sources/Sentry/SentryDebugImageProvider.m
Comment (brustolin) [reply Sources/Sentry/SentryDebugImageProvider.m]: Is the way I keep track of things I have done, even tho I haven't commit the fix.

File: Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/IO/SentryNSDataTrackerTests.swift]: >>and since swizzling is permanent, the reference to SentryNSDataTracker will be kept by the swizzled method.
>
> That's already the case.

Yes, and this is why `SentryNSDataTracker` needed to be singleton. To problem was in `SentryNSDataSwizzling`

> They shouldn't know that they are singletons...

You didn't get my point. 
SentryNSDataTracker used to follow the pattern of other classes that we already have. Classes that are used by Swizzling static classes.

But this doesn't matter anymore, I change this approach for SentryNSDataTracker

File: Sources/Sentry/SentryDebugImageProvider.m
Comment (philipphofmann) [reply Sources/Sentry/SentryDebugImageProvider.m]: Please commit the fix soon after resolving comments to avoid confusion.

File: Tests/SentryTests/Performance/SentryTracerTests.swift
Comment (brustolin) [reply Tests/SentryTests/Performance/SentryTracerTests.swift:175]: We already have tests for this in the `SentryNSDataTrackerTests` duplication seems unnecessary to me.

**Comment by brustolin** [comment]
> How can we ensure that buildStacktraceForCurrentThreadAsyncUnsafe works correctly on different OS versions and devices and doesn't crash?

Based on this: 
https://github.com/getsentry/sentry-cocoa/blob/c6504daa8453cd649bfa28238868b0680f347e58/Sources/SentryCrash/Recording/Tools/SentryCrashDynamicLinker.h?plain=1#L101
I belive that `dladdr` is meant to be used for this purpose. 
`sentrycrashdl_dladdr` has even the same signature. I would be more concern of using this handmade version than the OS native function. 

> One approach would be to add a UI test for running on real devices.
We have this already. iOS-Swift sample has plenty of file IO operation in the main thread when the app starts, and we have a lot of Ui tests with this sample that triggers the new code. At least is not crashing the app for sure.

File: Sources/Sentry/include/SentryThreadInspector.h
Comment (brustolin) [reply Sources/Sentry/include/SentryThreadInspector.h:17]: Do you remember any other function that uses comment reference so a can take a look?

**Comment by brustolin** [comment]
> I think a review from a native dev for the small part of C code would be great as suggested by the PR template:

Help from native dev are always welcome, but I dont think this PR touched any code we are not familiar with.

Please, If you have someone in mind who you think can help, don't hesitate to ping them directly, I don't mind.

File: Sources/Sentry/include/SentryThreadInspector.h
Comment (philipphofmann) [reply Sources/Sentry/include/SentryThreadInspector.h:17]: Sorry no. I would just remove the docs.

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (philipphofmann) [reply Sources/Sentry/SentryStacktraceBuilder.m]: @brustolin, please have a look at this vital comment.

**Review by philipphofmann — COMMENTED** [review]
> We have this already. iOS-Swift sample has plenty of file IO operation in the main thread when the app starts, and we have a lot of Ui tests with this sample that triggers the new code. At least is not crashing the app for sure.

Fine by me.

We still need to address

- [ ] https://github.com/getsentry/sentry-cocoa/pull/2573#discussion_r1069534968
- [ ] https://github.com/getsentry/sentry-cocoa/pull/2573#discussion_r1069522247

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m]: Sorry, somehow I miss this.

File: Sources/Sentry/SentryThreadInspector.m
Comment (brustolin) [inline Sources/Sentry/SentryThreadInspector.m:33]:
```suggestion
```

File: Sources/Sentry/include/SentryThreadInspector.h
Comment (brustolin) [inline Sources/Sentry/include/SentryThreadInspector.h:17]:
```suggestion
```

**Comment by brustolin** [comment]
Now we have more changes to C code.

@armcknight can you review my C changes?
Maybe @markushi would like to give it a try and review our C code?

**Review by philipphofmann — APPROVED** [review]
LGTM, we still need a review from a native dev, though.

**Review by markushi — APPROVED** [review]
I mainly checked the parts within `SentryCrash/Recording/**`, looks good to me!

File: Sources/Sentry/SentryStacktraceBuilder.m
Comment (brustolin) [reply Sources/Sentry/SentryStacktraceBuilder.m:109]: We want it `false` (which is set by the reset function) for `buildStacktraceForCurrentThread`, so I assume we dont need it.

File: Sources/SentryCrash/Recording/Tools/SentryCrashStackCursor.c
Comment (brustolin) [inline Sources/SentryCrash/Recording/Tools/SentryCrashStackCursor.c:53]: This is what worries me the most.
Im not sure whether `sentrycrashsc_resetCursor` is called in every scenario.
If `uselDlAddr` is no initialized to false during crashes, the value will be random and 'dladdr` may be used for crashes which is not desired.

Comment (philipphofmann) [reply Sources/SentryCrash/Recording/Tools/SentryCrashStackCursor.c:53]: I opened a PR to refactor this and remove this extra property https://github.com/getsentry/sentry-cocoa/pull/2659.

Comment (brustolin) [reply Sources/SentryCrash/Recording/Tools/SentryCrashStackCursor.c]: Perfect! This solution is definitely safer.

**Pull request merged**
